kphp-inspector: add strict search

i582 opened the pull request:
Now, if a class or function begins with a slash, then we treat the name as FQN and run strict search.

Strict search means that we are searching strictly by FQN.

The first thing we do is turn the FQN into the filename that the FQN is supposed to be in.

For example:
```
\VK\Namespace\SomeClass::__construct
// to
VK@Namespace@SomeClass@@__construct
```

After that, we limit this name with a dot on the right, thus we will only search for paths with this name.

```
VK@Namespace@SomeClass@@__construct.
```

And then we just start the file search, as in the current non-strict implementation.

Also added the `--debug` flag, more intended for embedding with the logging system for tools using kphp-inspector.

i582 committed:
added strict mode support

i582 committed:
fixed search for files

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+    return files.filter(file => file.endsWith('.h'));
+  }
+
+  let { qParts, longestQ } = queryToParts(query);
```
`[qParts, longestQ]` is above, but here you use `{}` destructuring.
I'd suggest to provide the `queryToParts` a better typing than it has now.

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+
+/**
+ * @param {string} query
+ * @returns {(string[]|string)[]}
```
See above about better typing.
+ A comment with an example input/return will be useful for understanding

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+
   let longestQLen = Math.max(...qParts.map(p => p.length));
   let longestQ = qParts.find(p => p.length === longestQLen);
+  utils.debug(`Longest query part: ${longestQ}`);
```
This debug seems too verbose

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+  }
+
+  let { qParts, longestQ } = queryToParts(query);
+  let files = findClassFilesWithNameLike(longestQ);
```
Here you use findClassFilesWithNameLike, but above (in strict) you use findFilesWithNameLike. (this issue doesn't exist in function searching)

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+ * @return {string[]}
+ */
+function findClassFilesWithNameLike(query) {
+  return findFilesWithNameLike(query, '/cl');
```
findFilesWithNameLike accepts a boolean, you provide a strring

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+ * @return {string[]}
+ */
+function findFilesWithNameLike(query, searchInNames = true, pathSuffix = '') {
+  const command = `find ${env.RUN_ARGV.KPHP_COMPILED_ROOT}${pathSuffix} ${searchInNames ? "-iname" : "-iwholename"} "*${query}*"`;
```
The name `searchInNames` doesn't describe what this parameter actually does

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
+  if (strict) {
+    // Convert the name to the form that KPHP uses for files.
+    const className = convertFqnToCppFileName(query);
+    const classnameWithDot = className + '\\.';
```
Here you use camelCase, above you use snake_case

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,139 @@ function doesFileContainFunction(fullFileName) {
-  if (!qParts.length) {
-    return [];
+function performSearchForFunction(query) {
+  const strict = query.startsWith('\\');
```
Maybe, here and below we shouldn't use a term 'strict', describing that "strict is fqn"? Let's just use "fqn" / "is_fqn" / etc.

i582 replied:
Fixed.

i582 committed:
fixed issues

- added comments
- refactored logic in findFilesWithNameLike

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,164 @@ function doesFileContainFunction(fullFileName) {
+
+  if (isFQN) {
+    const className = convertFqnToKphpStyleCppFileName(query);
+    const classNameWithDot = className + '\\.';
```
filenameWithDot?

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,164 @@ function doesFileContainFunction(fullFileName) {
+ * Searches for files for classes.
+ * @see findFilesWithNameLike
+ *
+ * @param {string} query Search string from console
```
It's not query, the comment is wrong

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,164 @@ function doesFileContainFunction(fullFileName) {
+ * Searches for files for functions with additional filtering.
+ * @see findFilesWithNameLike
+ *
+ * @param {string} query         Search string from console
```
It's not query, the comment is wrong

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,164 @@ function doesFileContainFunction(fullFileName) {
+    findQuery = `${query}*`;
+  } else {
+    findQuery = `*${query}*`;
+  }
```
Could be rewritten with a ternary right in declaration

i582 replied:
Fixed.

i582 committed:
fixed naming

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,159 @@ function doesFileContainFunction(fullFileName) {
-  if (!qParts.length) {
-    return [];
+function queryToParts(query) {
+  let queryParts = splitUserSearchStr(query);
```
Now, `splitUserSearchStr()` is being used only here and therefore seems a cognitive complication.
What do you think of inlining it here?

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,159 @@ function doesFileContainFunction(fullFileName) {
+
+  utils.debug(`Query parts: ${queryParts.join(', ')}`);
+
+  let longestQueryLen = Math.max(...queryParts.map(p => p.length));
```
longestPartLen

i582 replied:
Fixed.

unserialize commented on kphp-inspector/src/file_searcher.js:
```diff
@@ -38,42 +39,159 @@ function doesFileContainFunction(fullFileName) {
+ * @param {boolean} searchPrefix Specifies whether to search for a prefix or a substring
+ * @return {string[]}            An array of found files
+ */
+function findFilesWithNameLike(name, searchFolder = env.RUN_ARGV.KPHP_COMPILED_ROOT, searchPrefix = false) {
```
Default arguments are useless?

i582 replied:
Fixed.

i582 committed:
fixed issues

- inlined splitUserSearchStr
- fixed naming
- removed useless default value for findFilesWithNameLike argumenet

i582 committed:
fixed comment for findFilesWithNameLike
